### DISCUSSION TITLE
too stupid to use OMS USB sticks

### DISCUSSION THREAD

**Discussion by RobertTheLink** [wmbusmeters/wmbusmeters#852] [Q&A] [answered]
Hi all,
I have compiled the wmbusmeters on a raspi and got my good old  RTLSDR dongle to work .
Now I accidently could get hold on 3 different USB dongles,  im871a, Michael Rac and Amber.
However none of these is getting detected. So  I setup the docker version on my desktop, but still the USB dongles are not recognized.
Here the debug log when plugging in the im871a:
```console
(main) device /dev/ttyUSB0 not currently used, detect contents...
(serial) check if /dev/ttyUSB0 can be accessed
(serial) tty /dev/ttyUSB0 can be accessed
(serial) EVENT thread interrupted
(serial) before config: /dev/ttyUSB0 speed(B115200) input(ICRNL|IXON) output(ONLCR|OPOST) control(CLOCAL|CREAD|CSIZE|HUPCL) local(ECHO|ECHOCTL|ECHOE|ECHOK|ECHOKE|ICANON|IEXTEN|ISIG) special_chars(3,28,127,21,4,1,17,19,26,18,15,23,22,136,25,234,99,210,128,2,43) modem(DTR|RTS)
(serial) after config:  /dev/ttyUSB0 speed(B9600) input() output() control(CLOCAL|CREAD|CSIZE) local() special_chars(136,25,234,99,16) modem(DTR|RTS)
(serialtty) opened /dev/ttyUSB0 fd 5 (detect amb8465/amb3665)
(serial) received binary ""
(amb8465/3665) sending 6 bytes attempt 0
(serial /dev/ttyUSB0) sent "FF0A02008077"
(amb8465/3665) sent 6 bytes OK
(serial) EVENT thread interrupted
(amb8465/3665) reading response... 1
(serial) received binary ""
(amb8465/3665) reading response... 2
(serial) received binary ""
(amb8465/3665) reading response... 3
(serial) received binary ""
(amb8465/3665) are you there? no.
(serialtty) closed /dev/ttyUSB0 (detect amb8465/amb3665)
(serial) EVENT thread interrupted
(serial) before config: /dev/ttyUSB0 speed(B9600) input() output() control(CLOCAL|CREAD|CSIZE) local() special_chars(172,32,234,99,210,128,2,43) modem(DTR|RTS)
(serial) after config:  /dev/ttyUSB0 speed(B57600) input() output() control(CLOCAL|CREAD|CSIZE) local() special_chars(172,32,234,99,16) modem(DTR|RTS)
(serialtty) opened /dev/ttyUSB0 fd 5 (detect im871a)
(serial) received binary ""
(serial /dev/ttyUSB0) sent "A5010F00"
(serial) EVENT thread interrupted
(serial) received binary ""
(im871a/im170a) are you there? no.
(serialtty) closed /dev/ttyUSB0 (detect im871a)
(serial) EVENT thread interrupted
(serial) before config: /dev/ttyUSB0 speed(B57600) input() output() control(CLOCAL|CREAD|CSIZE) local() special_chars(172,32,234,99,210,128,2,43) modem(DTR|RTS)
(serial) after config:  /dev/ttyUSB0 speed(B19200) input() output() control(CLOCAL|CREAD|CSIZE) local() special_chars(172,32,234,99,16) modem(DTR|RTS)
(serialtty) opened /dev/ttyUSB0 fd 5 (detect rc1180)
(serial /dev/ttyUSB0) sent "00"
(serial) EVENT thread interrupted
(serial) received binary ""
(serial /dev/ttyUSB0) sent "30"
(serial) EVENT thread interrupted
(serial) received binary ""
(serialtty) closed /dev/ttyUSB0 (detect rc1180)
(serial) EVENT thread interrupted
(rc1180) are you there? no.
(serial) EVENT thread interrupted
(serial) before config: /dev/ttyUSB0 speed(B19200) input() output() control(CLOCAL|CREAD|CSIZE) local() special_chars(172,32,234,99,210,128,2,43) modem(DTR|RTS)
(serial) after config:  /dev/ttyUSB0 speed(B38400) input() output() control(CLOCAL|CREAD|CSIZE) local() special_chars(172,32,234,99,16) modem(DTR|RTS)
(serialtty) opened /dev/ttyUSB0 fd 5 (detect cul)
(serial /dev/ttyUSB0) sent "560D0A"
(serial) EVENT thread interrupted
(serial) received binary ""
(cul) probe response ""
(serial /dev/ttyUSB0) sent "560D0A"
(serial) EVENT thread interrupted
(serial) received binary ""
(cul) probe response ""
(serial /dev/ttyUSB0) sent "560D0A"
(serial) EVENT thread interrupted
(serial) received binary ""
(cul) probe response ""
(serialtty) closed /dev/ttyUSB0 (detect cul)
(serial) EVENT thread interrupted
(cul) are you there? no
(serial) EVENT thread interrupted
(serial) before config: /dev/ttyUSB0 speed(B38400) input() output() control(CLOCAL|CREAD|CSIZE) local() special_chars(172,32,234,99,210,128,2,43) modem(DTR|RTS)
(serial) after config:  /dev/ttyUSB0 speed(B115200) input() output() control(CLOCAL|CREAD|CSIZE) local() special_chars(172,32,234,99,16) modem(DTR|RTS)
(serialtty) opened /dev/ttyUSB0 fd 5 (detect iu880b)
(serial) received binary ""
(serial /dev/ttyUSB0) sent "C0C0C0C0C0C0C0C0C0C0C0C0C0C0C0C0C0C0C0C0C0C0C0C0C0C0C0C0C0C0"
(serial /dev/ttyUSB0) sent "C001030424C0"
(serial) EVENT thread interrupted
(serial) received binary ""
(iu880b) are you there? no.
(serialtty) closed /dev/ttyUSB0 (detect iu880b)
(serial) EVENT thread interrupted
(main) ignoring /dev/ttyUSB0, it does not respond as any of the supported wmbus devices.

```

my docker-compose.yml:
```yaml
version: "3.6"
services:
  wmbusmeters:
    image: weetmuts/wmbusmeters
    container_name: wmbusmeters
    restart: always
    volumes:
      - ./volumes/wmbusmeters:/wmbusmeters_data
      - ./volumes/wmbusmeters
      - /etc/localtime:/etc/localtime:ro
    devices:
      - /dev/:/dev/

```

and my wmbusmeter.conf:
```console
loglevel=debug
device=auto:t1
donotprobe=/dev/ttyAMA0
logtelegrams=false
format=json
meterfiles=/wmbusmeters_data/logs/meter_readings
meterfilesaction=overwrite
logfile=/wmbusmeters_data/logs/wmbusmeters.log
```
I have the strange feeling that there is a very easy solution to the issue, I am just missing it. Any idea?

**Comment by BIBOLV** [comment]
Hi,
By desktop you mean windows desktop? If so, I doubt that it would work.
I would suggest to get back to raspberry, plug im871a in, check in system logs that usb device is recognized, change wmbusmeters config to `device=/dev/ttyUSB0:im871a:c1,t1` and see if it works.

**Reply by RobertTheLink** [reply]
Hey @BIBOLV, thx for your answer!
I should have mentioned that "desktop"  is a Linux Mint machine,  errors are exactly the same as on the raspi.
Actually I tried already specifying a device as you mentioned. Changing wmbusmeters.conf, restarting the container, and  plugging in the im871a syslog shows:
``` console
Feb 17 22:23:37 myMedion kernel: [  773.033573] usb 3-1: new full-speed USB device number 6 using uhci_hcd
Feb 17 22:23:37 myMedion kernel: [  773.221271] usb 3-1: New USB device found, idVendor=10c4, idProduct=ea60, bcdDevice= 1.00
Feb 17 22:23:37 myMedion kernel: [  773.221274] usb 3-1: New USB device strings: Mfr=1, Product=2, SerialNumber=3
Feb 17 22:23:37 myMedion kernel: [  773.221277] usb 3-1: Product: CP2104 USB to UART Bridge Controller
Feb 17 22:23:37 myMedion kernel: [  773.221279] usb 3-1: Manufacturer: Silicon Labs
Feb 17 22:23:37 myMedion kernel: [  773.221281] usb 3-1: SerialNumber: 01E28B13
Feb 17 22:23:37 myMedion kernel: [  773.227278] cp210x 3-1:1.0: cp210x converter detected
Feb 17 22:23:37 myMedion kernel: [  773.233312] usb 3-1: cp210x converter now attached to ttyUSB0
Feb 17 22:23:37 myMedion mtp-probe: checking bus 3, device 6: "/sys/devices/pci0000:00/0000:00:1a.0/usb3/3-1"
Feb 17 22:23:37 myMedion mtp-probe: bus: 3, device: 6 was not an MTP device
```

and wmbusmeters.log shows:

```  console
(lookup) with file/hex "/dev/ttyUSB0" im871a
(serial) check if /dev/ttyUSB0 can be accessed
(serial) tty /dev/ttyUSB0 can be accessed
(serial) EVENT thread interrupted
(serialtty) could not open /dev/ttyUSB0 with 57600 baud N81
(im871a) could not open tty /dev/ttyUSB0 for detection
Expected im871a on /dev/ttyUSB0 but did not find it! Ignoring tty!
(verbose) ignoring device t1,c1
```

(serial) seems to get interrupted and then the device can no longer be accessed ?

**Comment by BIBOLV** [comment]
For cases like this, where debugging is needed, I would suggest to make standalone installation. Try running wmbusmeters from root on cmd `wmbusmeters --donotprobe=/dev/ttyAMA0 auto:t1,c1`

**Reply by RobertTheLink** [reply]
I tried `wmbusmeters --donotprobe=/dev/ttyAMA0 auto:t1,c1` and get
```console
╭─pi@raspi-03 ~/src
╰─$ wmbusmeters  --donotprobe=/dev/ttyAMA0 auto:t1,c1
No wmbus device detected, waiting for a device to be plugged in.
No meters configured. Printing id:s of all telegrams heard!
```

When I add the `--debug` option the output is exactly as in my first log post above. I.e. all possible dongles are checked, none is detected. And this is the same for any of the 3 dongles. 
Just to be sure, I tried my rtlsdr again: this is still working as expected.

**Comment by BIBOLV** [comment]
Did you updated wmbusmeters to latest version? Do you have access to windows - can you download imst app to confirm that dongle is actually working and if it has latest fw?

**Reply by RobertTheLink** [reply]
The Imst, Amber and Rac dongles are all working with their native Windows SW.
The Imst dongle has Firmware 1.5 from 05.01.2016, so this sounds pretty old.  As all the 3 dongles are borrowed, I should however not mess around with dongle Firmware, Unfortunately this is no option.
Updating from: `wmbusmeters: 1.11.0-67-g7f0ff31` to `wmbusmeters: 1.11.0-117-g231826a` did not change the behaviour, still the same messages in the log output.

**Comment by BIBOLV** [comment]
Strange! Check again that dongle is under ttyusb0 and try again cmd `./wmbusmeters --donotprobe=/dev/ttyAMA0 --debug /dev/ttyUSB0:im871a:c1`, but from root user, previously you executed under `pi`.

**Reply by RobertTheLink** [reply]
`/dev/ttyUSB0` is there when the dongle is plugged in.
executing as root did not cure the issue:
```console
╭─root@raspi-03 ~
╰─# ls /dev/ttyU*
/dev/ttyUSB0
╭─root@raspi-03 ~
╰─# wmbusmeters --donotprobe=/dev/ttyAMA0 --debug /dev/ttyUSB0:im871a:c1,t1
(wmbusmeters) version: _
(config) using device: /dev/ttyUSB0:im871a:t1,c1
(config) number of meters: 0
(lookup) with file/hex "/dev/ttyUSB0" im871a
(serial) check if /dev/ttyUSB0 can be accessed
(serial) tty /dev/ttyUSB0 can be accessed
(serial) EVENT thread interrupted
(serial) before config: /dev/ttyUSB0 speed(B57600) input() output() control(CLOCAL|CREAD|CSIZE) local() special_chars() modem(DTR|RTS)
(serial) after config:  /dev/ttyUSB0 speed(B57600) input() output() control(CLOCAL|CREAD|CSIZE) local() special_chars() modem(DTR|RTS)
(serialtty) opened /dev/ttyUSB0 fd 3 (detect im871a)
(serial) received binary ""
(serial /dev/ttyUSB0) sent "A5010F00"
(serial) EVENT thread interrupted
(serial) received binary ""
(im871a/im170a) are you there? no.
(serialtty) closed /dev/ttyUSB0 (detect im871a)
Expected im871a on /dev/ttyUSB0 but did not find it! Ignoring tty!
(serial) EVENT thread interrupted
(verbose) ignoring device t1,c1
No wmbus device detected, waiting for a device to be plugged in.
(shell) exec (capture output) "/bin/pidof"
(shell) arg "wmbusmetersd"
(shell) output: >>>4581
<<<
(serial) EVENT thread interrupted
(shell) return code 0
(shell) 10069 exited
Notice! Wmbusmeters daemon (pid 4581) is running and it might hog any wmbus devices.
(shell) exec (capture output) "/bin/pidof"
(shell) arg "wmbusmeters"
(shell) output: >>>10066
<<<
(serial) EVENT thread interrupted
(shell) return code 0
(shell) 10070 exited
(serial) registered regular callback HOT_PLUG_DETECTOR(0) every 2 seconds
No meters configured. Printing id:s of all telegrams heard!
(serial) waiting for stop
(serial) EVENT thread interrupted
(serial) TIMER thread interrupted```

**Comment by BIBOLV** [comment]
`Notice! Wmbusmeters daemon (pid 4581) is running and it might hog any wmbus devices.` - stop all possible processes that could lock dongle.

**Reply by RobertTheLink** [reply]
Hmmm,
there do no seem to be processes using ` /dev/ttyUSB0` except `wmbusmeters.d` for a short while  (which has PID 512 in this case). I run the same cmd a few times  before before and after plugging in the im781a:
```console
╭─pi@raspi-03 ~
╰─$ sudo fuser  /dev/ttyUSB0                                                                                                                                         1 ↵
Specified filename /dev/ttyUSB0 does not exist.
╭─pi@raspi-03 ~
╰─$ sudo fuser  /dev/ttyUSB0                                                                                                                                         1 ↵
/dev/ttyUSB0:          512
╭─pi@raspi-03 ~
╰─$ sudo fuser  /dev/ttyUSB0
/dev/ttyUSB0:          512
╭─pi@raspi-03 ~
╰─$ sudo fuser  /dev/ttyUSB0
╭─pi@raspi-03 ~ 
```

**Comment by BIBOLV** [comment]
You might have old udev rule that would call wmbusmeters service when usb device plugin is detected. You can remove it, disable service, reboot and try again as root.

**Reply by RobertTheLink** [reply]
Thank's again for all your effort!
if I search for dev rules for USB0, I do not find any (more) as I moved two files out of the rules dir. So this does not yield anything anymore: `grep -r 'USB0' /{{,local/}lib,run,etc}/udev/rules.d/`. Also no results when searching for wmbus instead. 
However the behaviour of the system is not changed.
I am using actually a fresh raspy Pi OS light with only zsh added besides wmbusmeters

**Comment by BIBOLV** [comment]
I wanted to suggest to make clean install - cool that you did it already. You compiled manually wmbusmeters or installed from pkg? Service is started when dongle is plugged in on this clean install?

**Reply by RobertTheLink** [reply]
Setting up the raps I enable ssh and serial interfaces. I compiled wmbusmeters manually. On a clean raspy light system git is obviously needed on top  to get the code, and then librtlsdr-dev and libncurses-dev in order to satisfy dependencies of  the make script.
dmesg is  identifying a new usb device when plugging in the im871a:
```console
[  316.729967] usb 1-1.2: new full-speed USB device number 4 using dwc_otg
[  316.877171] usb 1-1.2: New USB device found, idVendor=10c4, idProduct=ea60, bcdDevice= 1.00
[  316.877208] usb 1-1.2: New USB device strings: Mfr=1, Product=2, SerialNumber=3
[  316.877229] usb 1-1.2: Product: CP2104 USB to UART Bridge Controller
[  316.877246] usb 1-1.2: Manufacturer: Silicon Labs
[  316.877263] usb 1-1.2: SerialNumber: 01E28B13
[  316.970500] usbcore: registered new interface driver usbserial_generic
[  316.970578] usbserial: USB Serial support registered for generic
[  316.982806] usbcore: registered new interface driver cp210x
[  316.982884] usbserial: USB Serial support registered for cp210x
[  316.983058] cp210x 1-1.2:1.0: cp210x converter detected
[  316.987313] usb 1-1.2: cp210x converter now attached to ttyUSB0
```
Then I start wmbusmeters manually,  i.e. wmbusmeters is not yet enabled as service via  systemctl  at this point.

Do you expect the dongle to be identified as im871a in dmesg ?

**Reply by BIBOLV** [reply]
No, im871a will not be mentioned in dmesg. So now checking that no process is using ttyUSB0 and running cmd you still get same response?
Then try running with rtlsdr dongle just in case. If it works then I'm out of ideas and I would say that there is something wrong with imst dongle. For amber dongle it could related to wrong baud rate, but IMO it can't be changed on imst, but can`t verify it as I do not have one at my hand at the moment.

**Comment by RobertTheLink** [comment]
For the Amber stick I could solve the riddle: On the suppliers website we-online.com there is only one version of the stick listed, the AMB8665-M. However there is an other version available too, the AMB8665-AT2. This is the one I had borrowed. The -AT2 you need when using the Analyser SW and it is way more expensive than the -M, and apparently uses a different interface.

**Reply by BIBOLV** [reply]
imst actually also [have](https://shop.imst.de/wireless-modules/usb-radio-products/14/pa-im871a-wireless-m-bus-analyzer) specific dongles for debugging that are more expensive and maybe have some different interfaces, maybe from imst you also have one of those?

**Comment by weetmuts** [comment, marked as answer]
FYI the latest version of Würth/Amber stick is called Metis-II (which is the successor to amb8665-m and amb8465-m.)
I have tested the Metis-II USB dongle with sma connector, and it works fine with wmbusmeters. The model numbering is indeed a bit confusing but I have updated the README to reflect this.
